File: mysql-async/src/test/scala/com/github/mauricio/async/db/mysql/codec/MySQLFrameDecoderSpec.scala
/*
 * Copyright 2013 Maurício Linhares
 *
 * Maurício Linhares licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.github.mauricio.async.db.mysql.codec

import io.netty.buffer.ByteBuf
import io.netty.util.CharsetUtil
import com.github.mauricio.async.db.mysql.message.server._
import com.github.mauricio.async.db.util.ByteBufferUtils
import com.github.mauricio.async.db.util.ChannelWrapper.bufferToWrapper
import org.specs2.mutable.Specification
import com.github.mauricio.async.db.mysql.message.server.OkMessage
import com.github.mauricio.async.db.mysql.column.ColumnTypes
import io.netty.channel.embedded.EmbeddedChannel

class MySQLFrameDecoderSpec extends Specification {

  final val charset = CharsetUtil.UTF_8

  "decoder" should {

    "decode an OK message correctly" in {

      val buffer = createOkPacket()

      val decoder = this.createPipeline()

      decoder.writeInbound(buffer)

      val ok = decoder.readInbound().asInstanceOf[OkMessage]
      ok.affectedRows === 10
      ok.lastInsertId === 15
      ok.message === "this is a test"
      ok.statusFlags === 5
      ok.warnings === 6
    }

    "decode an error message" in {

      val content = "this is the error message"

      val buffer = createErrorPacket(content)

      val decoder = createPipeline()

      decoder.writeInbound(buffer)

      val error = decoder.readInbound().asInstanceOf[ErrorMessage]

      error.errorCode === 27
      error.errorMessage === content
      error.sqlState === "HZAWAY"

    }

    "on a query process it should correctly send an OK" in {

      val decoder = new MySQLFrameDecoder(charset, "[mysql-connection]")
      decoder.hasDoneHandshake = true
      val embedder = new EmbeddedChannel(decoder)
      embedder.config.setAllocator(LittleEndianByteBufAllocator.INSTANCE)

      decoder.queryProcessStarted()

      decoder.isInQuery must beTrue
      decoder.processingColumns must beTrue

      val buffer = createOkPacket()

      embedder.writeInbound(buffer) must beTrue
      embedder.readInbound().asInstanceOf[OkMessage].message === "this is a test"

      decoder.isInQuery must beFalse
      decoder.processingColumns must beFalse
    }

    "on query process it should correctly send an error" in {

      val decoder = new MySQLFrameDecoder(charset, "[mysql-connection]")
      decoder.hasDoneHandshake = true
      val embedder = new EmbeddedChannel(decoder)
      embedder.config.setAllocator(LittleEndianByteBufAllocator.INSTANCE)

      decoder.queryProcessStarted()

      decoder.isInQuery must beTrue
      decoder.processingColumns must beTrue

      val content = "this is a crazy error"

      val buffer = createErrorPacket(content)

      embedder.writeInbound(buffer) must beTrue
      embedder.readInbound().asInstanceOf[ErrorMessage].errorMessage === content

      decoder.isInQuery must beFalse
      decoder.processingColumns must beFalse

    }

    "on query process it should correctly handle a result set" in {

      val decoder = new MySQLFrameDecoder(charset, "[mysql-connection]")
      decoder.hasDoneHandshake = true
      val embedder = new EmbeddedChannel(decoder)
      embedder.config.setAllocator(LittleEndianByteBufAllocator.INSTANCE)

      decoder.queryProcessStarted()

      decoder.totalColumns === 0

      val columnCountBuffer = ByteBufferUtils.packetBuffer()
      columnCountBuffer.writeLength(2)
      columnCountBuffer.writePacketLength()

      embedder.writeInbound(columnCountBuffer)

      decoder.totalColumns === 2

      val columnId = createColumnPacket("id", ColumnTypes.FIELD_TYPE_LONG)
      val columnName = createColumnPacket("name", ColumnTypes.FIELD_TYPE_VARCHAR)

      embedder.writeInbound(columnId)

      embedder.readInbound().asInstanceOf[ColumnDefinitionMessage].name === "id"

      decoder.processedColumns === 1

      embedder.writeInbound(columnName)

      embedder.readInbound().asInstanceOf[ColumnDefinitionMessage].name === "name"

      decoder.processedColumns === 2

      embedder.writeInbound(this.createEOFPacket())

      embedder.readInbound().asInstanceOf[ColumnProcessingFinishedMessage].eofMessage.flags === 8765

      decoder.processingColumns must beFalse

      val row = ByteBufferUtils.packetBuffer()
      row.writeLenghtEncodedString("1", charset)
      row.writeLenghtEncodedString("some name", charset)
      row.writePacketLength()

      embedder.writeInbound(row)

      embedder.readInbound().isInstanceOf[ResultSetRowMessage] must beTrue

      embedder.writeInbound(this.createEOFPacket())

      decoder.isInQuery must beFalse
    }

  }

  def createPipeline(): EmbeddedChannel = {
    val decoder = new MySQLFrameDecoder(charset, "[mysql-connection]")
    decoder.hasDoneHandshake = true
    val channel = new EmbeddedChannel(decoder)
    channel.config.setAllocator(LittleEndianByteBufAllocator.INSTANCE)
    channel
  }

  def createOkPacket() : ByteBuf = {
    val buffer = ByteBufferUtils.packetBuffer()
    buffer.writeByte(0)
    buffer.writeLength(10)
    buffer.writeLength(15)
    buffer.writeShort(5)
    buffer.writeShort(6)
    buffer.writeBytes("this is a test".getBytes(charset))
    buffer.writePacketLength()
    buffer
  }

  def createErrorPacket(content : String) : ByteBuf = {
    val buffer = ByteBufferUtils.packetBuffer()
    buffer.writeByte(0xff)
    buffer.writeShort(27)
    buffer.writeByte('H')
    buffer.writeBytes("ZAWAY".getBytes(charset))
    buffer.writeBytes(content.getBytes(charset))
    buffer.writePacketLength()
    buffer
  }

  def createColumnPacket( name : String, columnType : Int ) : ByteBuf = {
    val buffer = ByteBufferUtils.packetBuffer()
    buffer.writeLenghtEncodedString("def", charset)
    buffer.writeLenghtEncodedString("some_schema", charset)
    buffer.writeLenghtEncodedString("some_table", charset)
    buffer.writeLenghtEncodedString("some_table", charset)
    buffer.writeLenghtEncodedString(name, charset)
    buffer.writeLenghtEncodedString(name, charset)
    buffer.writeLength(12)
    buffer.writeShort(0x03)
    buffer.writeInt(10)
    buffer.writeByte(columnType)
    buffer.writeShort(76)
    buffer.writeByte(0)
    buffer.writeShort(56)
    buffer.writePacketLength()
    buffer
  }

  def createEOFPacket() : ByteBuf = {
    val buffer = ByteBufferUtils.packetBuffer()
    buffer.writeByte(0xfe)
    buffer.writeShort(879)
    buffer.writeShort(8765)

    buffer.writePacketLength()

    buffer
  }



}

File: mysql-async/src/test/java/com/github/jasync/sql/db/mysql/codec/MySQLFrameDecoderSpec.kt
package com.github.jasync.sql.db.mysql.codec

import com.github.jasync.sql.db.mysql.column.ColumnTypes
import com.github.jasync.sql.db.mysql.message.server.ColumnDefinitionMessage
import com.github.jasync.sql.db.mysql.message.server.ColumnProcessingFinishedMessage
import com.github.jasync.sql.db.mysql.message.server.ErrorMessage
import com.github.jasync.sql.db.mysql.message.server.OkMessage
import com.github.jasync.sql.db.mysql.message.server.ResultSetRowMessage
import com.github.jasync.sql.db.util.ByteBufferUtils
import com.github.jasync.sql.db.util.ChannelWrapper
import com.github.jasync.sql.db.util.writeLength
import com.github.jasync.sql.db.util.writeLengthEncodedString
import io.netty.buffer.ByteBuf
import io.netty.channel.embedded.EmbeddedChannel
import io.netty.util.CharsetUtil
import org.junit.Test
import java.nio.charset.Charset
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class MySQLFrameDecoderSpec {

    private val charset: Charset = CharsetUtil.UTF_8

    @Test
    fun `decode an OK message correctly`() {
        val buffer = createOkPacket()
        val decoder = this.createPipeline()
        decoder.writeInbound(buffer)

        val ok = decoder.readInbound() as OkMessage
        assertEquals(10, ok.affectedRows)
        assertEquals(15, ok.lastInsertId)
        assertEquals("this is a test", ok.message)
        assertEquals(5, ok.statusFlags)
        assertEquals(6, ok.warnings)
    }

    @Test
    fun `decode an error message`() {
        val content = "this is the error message"
        val buffer = createErrorPacket(content)
        val decoder = createPipeline()

        decoder.writeInbound(buffer)

        val error = decoder.readInbound() as ErrorMessage
        assertEquals(27, error.errorCode)
        assertEquals(content, error.errorMessage)
        assertEquals("HZAWAY", error.sqlState)
    }

    @Test
    fun `on a query process it should correctly send an OK`() {
        val decoder = MySQLFrameDecoder(charset, "[mysql-connection]")
        decoder.hasDoneHandshake = true
        val embedder = EmbeddedChannel(decoder)
        embedder.config().allocator = LittleEndianByteBufAllocator.INSTANCE

        decoder.queryProcessStarted()

        assertTrue(decoder.isInQuery)
        assertTrue(decoder.processingColumns)

        val buffer = createOkPacket()
        assertTrue(embedder.writeInbound(buffer))
        assertEquals("this is a test", (embedder.readInbound() as OkMessage).message)

        assertFalse(decoder.isInQuery)
        assertFalse(decoder.processingColumns)
    }

    @Test
    fun `on query process it should correctly send an error`() {

        val decoder = MySQLFrameDecoder(charset, "[mysql-connection]")
        decoder.hasDoneHandshake = true
        val embedder = EmbeddedChannel(decoder)
        embedder.config().allocator = LittleEndianByteBufAllocator.INSTANCE

        decoder.queryProcessStarted()

        assertTrue(decoder.isInQuery)
        assertTrue(decoder.processingColumns)

        val content = "this is a crazy error"

        val buffer = createErrorPacket(content)

        assertTrue(embedder.writeInbound(buffer))
        assertEquals(content, (embedder.readInbound() as ErrorMessage).errorMessage)
        assertFalse(decoder.isInQuery)
        assertFalse(decoder.processingColumns)
    }

    @Test
    fun `on query process it should correctly handle a result set`() {

        val decoder = MySQLFrameDecoder(charset, "[mysql-connection]")
        decoder.hasDoneHandshake = true
        val embedder = EmbeddedChannel(decoder)
        embedder.config().allocator = LittleEndianByteBufAllocator.INSTANCE

        decoder.queryProcessStarted()

        assertEquals(0L, decoder.totalColumns)

        val columnCountBuffer = ByteBufferUtils.packetBuffer()
        columnCountBuffer.writeLength(2)
        ChannelWrapper.writePacketLength(columnCountBuffer)

        embedder.writeInbound(columnCountBuffer)

        assertEquals(2, decoder.totalColumns)

        val columnId = createColumnPacket("id", ColumnTypes.FIELD_TYPE_LONG)
        val columnName = createColumnPacket("name", ColumnTypes.FIELD_TYPE_VARCHAR)

        embedder.writeInbound(columnId)
        assertEquals("id", (embedder.readInbound() as ColumnDefinitionMessage).name)
        assertEquals(1, decoder.processedColumns)

        embedder.writeInbound(columnName)

        assertEquals("name", (embedder.readInbound() as ColumnDefinitionMessage).name)

        assertEquals(2, decoder.processedColumns)

        embedder.writeInbound(this.createEOFPacket())

        assertEquals(8765, (embedder.readInbound() as ColumnProcessingFinishedMessage).eofMessage.flags)

        assertFalse(decoder.processingColumns)

        val row = ByteBufferUtils.packetBuffer()
        row.writeLengthEncodedString("1", charset)
        row.writeLengthEncodedString("some name", charset)
        ChannelWrapper.writePacketLength(row, 0)

        embedder.writeInbound(row)

        embedder.readInbound() as ResultSetRowMessage

        embedder.writeInbound(this.createEOFPacket())

        assertFalse(decoder.isInQuery)
    }

    fun createPipeline(): EmbeddedChannel {
        val decoder = MySQLFrameDecoder(charset, "[mysql-connection]")
        decoder.hasDoneHandshake = true
        val channel = EmbeddedChannel(decoder)
        channel.config().allocator = LittleEndianByteBufAllocator.INSTANCE
        return channel
    }

    fun createOkPacket(): ByteBuf {
        val buffer = ByteBufferUtils.packetBuffer()
        buffer.writeByte(0)
        buffer.writeLength(10)
        buffer.writeLength(15)
        buffer.writeShort(5)
        buffer.writeShort(6)
        buffer.writeBytes("this is a test".toByteArray())
        ChannelWrapper.writePacketLength(buffer, 0)
        return buffer
    }

    fun createErrorPacket(content: String): ByteBuf {
        val buffer = ByteBufferUtils.packetBuffer()
        buffer.writeByte(0xff)
        buffer.writeShort(27)
        buffer.writeByte('H'.toByte().toInt())
        buffer.writeBytes("ZAWAY".toByteArray(charset))
        buffer.writeBytes(content.toByteArray(charset))
        ChannelWrapper.writePacketLength(buffer, 0)
        return buffer
    }

    fun createColumnPacket(name: String, columnType: Int): ByteBuf {
        val buffer = assertNotNull(ByteBufferUtils.packetBuffer())
        buffer.writeLengthEncodedString("def", charset)
        buffer.writeLengthEncodedString("some_schema", charset)
        buffer.writeLengthEncodedString("some_table", charset)
        buffer.writeLengthEncodedString("some_table", charset)
        buffer.writeLengthEncodedString(name, charset)
        buffer.writeLengthEncodedString(name, charset)
        ChannelWrapper.writeLength(buffer, 12)
        buffer.writeShort(0x03)
        buffer.writeInt(10)
        buffer.writeByte(columnType)
        buffer.writeShort(76)
        buffer.writeByte(0)
        buffer.writeShort(56)
        ChannelWrapper.writePacketLength(buffer, 0)
        return buffer
    }

    fun createEOFPacket(): ByteBuf {
        val buffer = ByteBufferUtils.packetBuffer()
        buffer.writeByte(0xfe)
        buffer.writeShort(879)
        buffer.writeShort(8765)

        ByteBufferUtils.writePacketLength(buffer, 0)

        return buffer
    }
}
